sdk/protocol2.go

```diff
@@ -0,0 +1,164 @@
+package sdk
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/newrelic/infra-integrations-sdk/args"
+	"github.com/newrelic/infra-integrations-sdk/cache"
+	"github.com/newrelic/infra-integrations-sdk/log"
+	"github.com/newrelic/infra-integrations-sdk/metric"
+	"github.com/pkg/errors"
+)
+
+// Entity is the producer of the data. Entity could be a host, a container, a pod, or whatever unit of meaning.
+type Entity struct {
+	Name string `json:"name"`
+	Type string `json:"type"`
+}
+
+// EntityData defines all the data related to a particular event from an entity.
+type EntityData struct {
+	Entity    Entity             `json:"entity"`
+	Metrics   []metric.MetricSet `json:"metrics"`
+	Inventory Inventory          `json:"inventory"`
+	Events    []Event            `json:"events"`
+}
+
+// IntegrationProtocol2 defines the format of the output JSON that integrations will return for protocol 2.
+type IntegrationProtocol2 struct {
+	sync.Mutex
+	Name               string        `json:"name"`
+	ProtocolVersion    string        `json:"protocol_version"`
+	IntegrationVersion string        `json:"integration_version"`
+	Data               []*EntityData `json:"data"`
+	prettyOutput       bool
+	writer             io.Writer
+}
+
+// NewIntegrationProtocol2WithWriter initializes a new instance of the integration protocol 2 specifying a writer instead of using stdout by default.
+func NewIntegrationProtocol2WithWriter(name string, version string, arguments interface{}, writer io.Writer) (*IntegrationProtocol2, error) {
+	i, err := NewIntegrationProtocol2(name, version, arguments)
+	if err != nil {
+		return nil, err
+	}
+
+	i.writer = writer
+
+	return i, nil
+}
+
+// NewIntegrationProtocol2 initializes a new instance of the integration protocol 2.
+func NewIntegrationProtocol2(name string, version string, arguments interface{}) (*IntegrationProtocol2, error) {
+	err := args.SetupArgs(arguments)
+	if err != nil {
+		return nil, err
+	}
+	defaultArgs := args.GetDefaultArgs(arguments)
+
+	log.SetupLogging(defaultArgs.Verbose)
+
+	// Avoid working with an uninitialized or in error state cache
+	if err = cache.Status(); err != nil {
+		return nil, err
+	}
+
+	integration := &IntegrationProtocol2{
+		Name:               name,
+		ProtocolVersion:    "2",
+		IntegrationVersion: version,
+		prettyOutput:       defaultArgs.Pretty,
+		writer:             os.Stdout, // StdOut by default
+	}
+
+	return integration, nil
+}
+
+// Entity method creates or retrieves an already created EntityData.
+func (integration *IntegrationProtocol2) Entity(entityName, entityType string) (*EntityData, error) {
+	if entityName == "" || entityType == "" {
+		return nil, errors.New("entityName and entityType should be not empty")
+	}
+
+	integration.Lock()
```

Comment:
I assume this process is fast enough and not too concurrent so we can afford locking the whole method until it completes.

Reply:
As could be as concurrent as you want we must add this lock, which in some cases could be only Read lock or Write, but trying to simplify by now.

This is a must across all our sdk. In fact we should add it in some more places.